### PR TITLE
Deprecate impute_missing_data and TSK_IMPUTE_MISSING_DATA

### LINKED ISSUE #716: Some questions about missing data
Sorry to bring this up again, but #714 made @bhaller ask some good questions. Two observations:

1. It is surprising that `simplify()` can make a formerly non-missing genotype become missing.
2. An isolated node is considered to have a missing genotype even if it has a mutation above it, which is also surprising.

The first thing is defensible, because we treat the state of every root as the "ancestral state" even though there's got to be some element of unknown-ness. But it is certainly surprising, and makes me wonder if we can modify this behavior. The easiest way I could think of to modify it would be to *have some way to mark an isolated node as "not actually missing". *

The second thing does not feel right.

Here's a demonstration of (2), btw:
```
tables = tskit.TableCollection(sequence_length=1.0)
tables.nodes.add_row(time=0, flags=1)
tables.nodes.add_row(time=0, flags=1)
tables.nodes.add_row(time=0, flags=1)
tables.nodes.add_row(time=1, flags=0)
tables.edges.add_row(parent=3, child=1, left=0, right=1)
tables.edges.add_row(parent=3, child=2, left=0, right=1)
tables.sites.add_row(position=0.5, ancestral_state='x')
tables.mutations.add_row(site=0, node=0, derived_state='y')
ts = tables.tree_sequence()
ts.genotype_matrix()
# array([[-1,  0,  0]], dtype=int8)
```

I don't remember discussion of this point when working out missing-ness, but maybe I'm forgetting it?

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #716 
I have made the arguments to the functions that take the new argument all keyword-only. This didn't break any of our test code and will only break those users that are using the arguments positionally. While this break could be avoided, I think it is best as it allows us to remove the old argument eventually with out more subtle breakage.

### CODE REVIEW

File: python/tskit/trees.py [hunk 1]
```diff
@@ -3691,14 +3697,14 @@ def haplotypes(self, impute_missing_data=False, missing_data_character="-"):
...
-        derived  mutational state for that node). This was the default
-        behaviour in versions prior to 0.2.0.
+        If ``isolated_as_missing`` is True (the default), isolated samples will be
+        treated as :ref:`missing data<sec_data_model_missing_data>` and will be
+        represented in the string by the ``missing_data_character``. If
+        instead it is set to False, missing data will be assigned the ancestral state
+        (unless they have mutations directly above them, in which case they will take
+        the most recent derived mutational state for that node). This was the default
+        behaviour in versions prior to 0.2.0. Prior to 0.3.0 the `impute_missing_data`
+        argument controlled this behaviour.
```
Comment (petrelharp) [inline python/tskit/trees.py:3707]: Maybe
```
        If ``isolated_as_missing`` is True (the default), isolated samples without mutations directly above them
        will be treated as :ref:`missing data<sec_data_model_missing_data>` and will be
```

File: python/tskit/trees.py [hunk 2]
```diff
@@ -3713,22 +3719,37 @@ def haplotypes(self, impute_missing_data=False, missing_data_character="-"):
...
         :return: An iterator over the haplotype strings for the samples in
             this tree sequence.
-        :param bool impute_missing_data: If True, the allele assigned to any
-            isolated samples is the ancestral state; that is, we impute
-            missing data as the ancestral state. Default: False.
+        :param bool isolated_as_missing: If True, the allele assigned to
+            missing (isolated) samples is ``missing_data_character``. If False,
+            missing (isolated) samples will be assigned the ancestral state or
+            that of the mutation above them, if any.
+            Default: True.
```
Comment (petrelharp) [inline python/tskit/trees.py:3726]: how about
```
        :param bool isolated_as_missing: If True, the allele assigned to
            missing samples (i.e., isolated samples without mutations) is the ``missing_data_character``. If False,
            missing samples will be assigned the ancestral state.
            Default: True.
```

Comment (petrelharp) [reply python/tskit/trees.py:3726]: I guess I'm interpreting our position to be that samples with a mutation are *not* missing - do you agree, @hyanwong, @jeromekelleher?

File: python/tskit/trees.py [hunk 3]
```diff
@@ -3713,22 +3719,37 @@ def haplotypes(self, impute_missing_data=False, missing_data_character="-"):
...
         :rtype: collections.abc.Iterable
         :raises: TypeError if the ``missing_data_character`` or any of the alleles
             at a site or the are not a single ascii character.
         :raises: ValueError
             if the ``missing_data_character`` exists in one of the alleles
         """
+        if impute_missing_data is not None:
+            warnings.warn(
+                "The impute_missing_data parameter was deprecated in 0.3.0 and will"
+                " be removed. Use ``isolated_as_missing``, but inverting the value.",
```
Comment (petrelharp) [inline python/tskit/trees.py:3743]: perhaps "Use ``isolated_as_missing=False`` instead of ``impute_missing_data=True``."?

Comment (petrelharp) [reply python/tskit/trees.py:3743]: since "invert" the value might be confusing (which value?)

File: python/tskit/trees.py [hunk 4]
```diff
@@ -3713,22 +3719,37 @@ def haplotypes(self, impute_missing_data=False, missing_data_character="-"):
...
+                DeprecationWarning,
+            )
+        # Only use impute_missing_data if isolated_as_missing has the default value
+        if isolated_as_missing is None:
+            isolated_as_missing = not impute_missing_data
```
Comment (petrelharp) [inline python/tskit/trees.py:3748]: ah, `not None` is `True` =)

File: python/tskit/trees.py [hunk 5]
```diff
@@ -3785,23 +3812,24 @@ def variants(
...
-        using the ``alleles`` parameter or not). See the :class:`Variant`
-        class for more details on how missing data is reported.
-
-        Missing data is reported by default, but if ``impute_missing_data``
-        is set to to True, missing data will be imputed such that all
-        isolated samples are assigned the ancestral state (unless they have
-        mutations directly above them, in which case they will take the most
-        recent derived mutational state for that node). This was the
-        default behaviour in versions prior to 0.2.0.
+        If isolated samples are present at a given site, they will be
```
Comment (petrelharp) [inline python/tskit/trees.py:3818]: "If isolated samples are present at a given site without mutations above them,"

File: python/tskit/trees.py [hunk 6]
```diff
@@ -3785,23 +3812,24 @@ def variants(
...
+        (-1) to identify these missing samples, and the ``alleles`` tuple will
+        end with the value ``None`` (note that this is true whether we specify
+        a fixed mapping using the ``alleles`` parameter or not).
+        See the :class:`Variant` class for more details on how missing data is
+        reported.
+
+        Missing data is reported by default, but if ``isolated_as_missing``
+        is set to to False, isolated samples will be assigned the ancestral
+        state (unless they have mutations directly above them, in which case
+        they will take the most recent derived mutational state for that node).
```
Comment (petrelharp) [inline python/tskit/trees.py:3830]: how about
```

        Such samples are treated as missing data by default, but if ``isolated_as_missing``
        is set to to False, they will not be treated as missing, and so assigned the ancestral
        state.
```

File: python/tskit/trees.py [hunk 7]
```diff
@@ -3813,24 +3841,38 @@ class for more details on how missing data is reported.
             as a Python bytes object. Legacy use only.
         :param array_like samples: An array of sample IDs for which to generate
             genotypes, or None for all samples. Default: None.
-        :param bool impute_missing_data: If True, the allele assigned to any
-            isolated samples is the ancestral state; that is, we impute
-            missing data as the ancestral state. Default: False.
+        :param bool isolated_as_missing: If True, the allele assigned to
+            missing (isolated) samples is ``missing_data_character``. If False,
+            missing (isolated) samples will be assigned the ancestral state or
+            that of the mutation above them, if any.
```
Comment (petrelharp) [inline python/tskit/trees.py:3847]: as above

File: python/tskit/trees.py [hunk 8]
```diff
@@ -3846,40 +3888,61 @@ class for more details on how missing data is reported.
...
+        If isolated samples are present at a given site, they will be
+        interpreted as :ref:`missing data<sec_data_model_missing_data>`
+        the genotypes array will contain a special value :data:`MISSING_DATA`
+        (-1) to identify these missing samples.
+        Missing data is reported by default, but if ``isolated_as_missing``
+        is set to to False, isolated samples will be assigned the ancestral
+        state (unless they have mutations directly above them, in which case
+        they will take the most recent derived mutational state for that node).
+        This was the default behaviour in versions prior to 0.2.0. Prior to 0.3.0
+        the `impute_missing_data` argument controlled this behaviour.
```
Comment (petrelharp) [inline python/tskit/trees.py:3909]: as above

File: python/tskit/trees.py [hunk 9]
```diff
@@ -3846,40 +3888,61 @@ class for more details on how missing data is reported.
...
             access them sequentially using the :meth:`.variants` iterator.
 
-        :param bool impute_missing_data: If True, the allele assigned to any
-            isolated samples is the ancestral state; that is, we impute
-            missing data as the ancestral state. Default: False.
+        :param bool isolated_as_missing: If True, the allele assigned to
+            missing (isolated) samples is ``missing_data_character``. If False,
+            missing (isolated) samples will be assigned the ancestral state or
+            that of the mutation above them, if any.
+            Default: True.
```
Comment (petrelharp) [inline python/tskit/trees.py:3920]: as above

File: python/tskit/trees.py (hunk 2 above)
Comment (hyanwong) [reply python/tskit/trees.py:3726]: Yes, completely

File: python/tskit/trees.py (hunk 1 above)
Comment (benjeffery) [reply python/tskit/trees.py:3708]: Done

File: python/tskit/trees.py (hunk 2 above)
Comment (benjeffery) [reply python/tskit/trees.py:3727]: Done

File: python/tskit/trees.py (hunk 3 above)
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py (hunk 5 above)
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py (hunk 6 above)
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py (hunk 7 above)
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py (hunk 8 above)
Comment (benjeffery) [reply python/tskit/trees.py:3911]: Done

File: python/tskit/trees.py (hunk 9 above)
Comment (benjeffery) [reply python/tskit/trees.py:3922]: Done

File: c/tskit/genotypes.h [hunk 10]
```diff
@@ -33,7 +33,10 @@ extern "C" {
 #include <tskit/trees.h>
 
 #define TSK_16_BIT_GENOTYPES (1 << 0)
-#define TSK_IMPUTE_MISSING_DATA (1 << 1)
+#define TSK_ISOLATED_NOT_MISSING (1 << 1)
+
+/* This flag was deprected in 0.3.0 and replaced with TSK_ISOLATED_NOT_MISSING */
+#define TSK_IMPUTE_MISSING_DATA TSK_ISOLATED_NOT_MISSING
```
Comment (jeromekelleher) [inline c/tskit/genotypes.h:39]: Just nuke it - this is why we're still on 0.99, we reserve the right to break people's code.

File: python/tests/test_genotypes.py [hunk 11]
```diff
@@ -437,18 +437,31 @@ def test_zero_edge_missing_data(self):
...
+        self.assertTrue(np.all(Gm[0] == -1))
+        self.assertTrue(np.all(Gm[1] == -1))
+        self.assertTrue(np.all(Gm[-1] == -1))
+        self.assertTrue(np.all(Gm[-2] == -1))
+        Gm2 = ts.genotype_matrix(isolated_as_missing=True)
+        self.assertTrue(np.array_equal(Gm, Gm2))
+
+        # Test deprecated param
+        Gi = ts.genotype_matrix(impute_missing_data=True)
+        print(Gi)
```
Comment (jeromekelleher) [inline python/tests/test_genotypes.py:455]: Stray prints

File: python/tskit/trees.py [hunk 12]
```diff
@@ -3673,7 +3673,13 @@ def trees(
         )
         return TreeIterator(tree)
 
-    def haplotypes(self, impute_missing_data=False, missing_data_character="-"):
+    def haplotypes(
+        self,
+        *,
+        isolated_as_missing=None,
+        missing_data_character="-",
+        impute_missing_data=None,
```
Comment (jeromekelleher) [inline python/tskit/trees.py:3681]: I don't think ``impute_missing_data`` was ever released for haplotypes, so we can probably just nuke it. (best to check, though)

File: python/tskit/trees.py (hunk 2 above)
Comment (jeromekelleher) [reply python/tskit/trees.py:3727]: Yes, we've fixed that in one of the in-flight PRs.

File: python/tskit/trees.py [hunk 13]
```diff
@@ -3846,40 +3890,62 @@ class for more details on how missing data is reported.
                 genotypes = bytes_genotypes.tobytes()
             yield Variant(site, alleles, genotypes)
 
-    def genotype_matrix(self, impute_missing_data=False, alleles=None):
+    def genotype_matrix(
+        self, *, isolated_as_missing=None, alleles=None, impute_missing_data=None
```
Comment (jeromekelleher) [inline python/tskit/trees.py:3894]: I'm not sure ``impute_missing_data`` for the genotype matrix was released either, worth having a check.

**Review by jeromekelleher — COMMENTED** [review]
Looks great, thanks @benjeffery!

File: c/tskit/genotypes.h (hunk 10 above)
Comment (jeromekelleher) [reply c/tskit/genotypes.h:39]: Unless you object @petrelharp?

Comment (petrelharp) [reply c/tskit/genotypes.h:39]: no problems here

Comment (benjeffery) [reply c/tskit/genotypes.h]: Removed in https://github.com/tskit-dev/tskit/pull/794/commits/1cc1e9eb659bdc71f83d74d262a8a488f0875dda

Comment (benjeffery) [reply c/tskit/genotypes.h]: In retrospect was silly to refer to the python version number in the C....

File: python/tests/test_genotypes.py (hunk 11 above)
Comment (benjeffery) [reply python/tests/test_genotypes.py]: https://github.com/tskit-dev/tskit/pull/794/commits/af69ca95b50156f324527822bd5db769ef690771

File: python/tskit/trees.py (hunk 12 above)
Comment (benjeffery) [reply python/tskit/trees.py:3681]: From what I can see it was released: https://github.com/tskit-dev/tskit/blob/0.2.3/python/tskit/trees.py#L2830

File: python/tskit/trees.py (hunk 13 above)
Comment (benjeffery) [reply python/tskit/trees.py:3894]: This was in 0.2.3 also https://github.com/tskit-dev/tskit/blob/0.2.3/python/tskit/trees.py#L2936

File: python/tskit/trees.py (hunk 12 above)
Comment (jeromekelleher) [reply python/tskit/trees.py:3681]: Ah, my bad. Thanks.

**Comment by benjeffery** [comment]
Ok I've addressed comments and added the changelog and squashed. I think this is good to go.

**Comment by petrelharp** [comment]
Whoops - we should probably update the docs also to clarify about missingness? I'll have a go at this.

**Comment by benjeffery** [comment]
> Whoops - we should probably update the docs also to clarify about missingness? I'll have a go at this.

Great - glad to have your eyes on these last few things before release!

**Review by jeromekelleher — APPROVED** [review]
LGTM, thanks @benjeffery !

**Pull request merged**